### PR TITLE
Js docx nv 1697

### PR DESCRIPTION
**Pull request by jenniferstruthers1-nhs**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,3 +39,19 @@ Includes learning on:
     - Recap of workshop 2
     - Functions
     - Importing files
+
+**Workshop 4: Intro to PySpark**
```
Comment (warren-davies4) [inline README.md:43]: thanks, sorry I forgot to add this!

File: workshop_5/docx.ipynb [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+   "execution_count": 57,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def replace_text_in_paragraph(paragraph, replacements):\n",
+    "    \"\"\"\n",
+    "    Replace text in a paragraph based on a dictionary of replacements.\n",
+    "    \"\"\"\n",
+    "    for old_text, new_text in replacements.items():\n",
+    "        if old_text in paragraph.text:\n",
```
Comment (warren-davies4) [inline workshop_5/docx.ipynb]: here we could reduce nesting and intruduce `continue` with a guard clause:

```py
if old_text not in paragraph.text:
  continue

for run in...
```

File: workshop_5/docx.ipynb [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+   "outputs": [],
+   "source": [
+    "def generate_documents_from_spark_df(spark_df, template_docx, output_folder, placeholder_mapping):\n",
+    "    # Collect the DataFrame rows into a list of dictionaries\n",
+    "    rows = spark_df.collect()\n",
+    "    \n",
+    "    # Iterate through each row in the DataFrame\n",
+    "    for row in rows:\n",
+    "        # Load the template document\n",
+    "        doc = Document(template_docx)\n",
```
Comment (warren-davies4) [inline workshop_5/docx.ipynb]: I think this declaration could be moved above the loop

**Review by warren-davies4 — CHANGES_REQUESTED** [review]
This looks great Jen! I like how the session moves to using more complex examples that make use of the things learned in the previous ones, functions, loops and so on.

I made a couple of suggestions.

I think we've only touched on dictionaries so far, I wonder whether it's worth having a a short notebook before this one to more formally introduce them, and .items()/.values() etc. Or, we could just explain as we go. Just an idea - your call!

**Comment by jenniferstruthers1-nhs** [comment]
@warren-davies4 
I added a notebook on dictionaries, I think it's a good call as it might be a bit confusing to jump into them while learning something else new at the same time. It's fairly short but I think covers anything that would be necessary.

I didn't create any homework, but I think since we're supposed to be starting the thin slice that that will be enough

File: workshop_5/dictionaries.ipynb [hunk 4]
```diff
@@ -0,0 +1,300 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Intro to dictionaries\n",
+    "\n",
+    "A dictionary in Python is an unordered collection of items. This is different from a list, where you can access the first, second, nth item. You cannot do this with dictionaries. Instead, each item is a key-value pair, where each key is unique and this ishow you would access the 'value'.\n",
```
Comment (warren-davies4) [inline workshop_5/dictionaries.ipynb]:
```suggestion
    "A dictionary in Python is an ordered collection of items. This is different from a list, where you can access the first, second, nth item. You cannot do this with dictionaries. Instead, each item is a key-value pair, where each key is unique and this is how you would access the 'value'.\n",
```

File: workshop_5/dictionaries.ipynb [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
```
Comment (warren-davies4) [inline workshop_5/dictionaries.ipynb]: did you mean to leave the outputs in?

**Review by warren-davies4 — APPROVED** [review]
Great stuff, just fixed a typo or two

**Pull request merged**